### DISCUSSION TITLE
Appy Duty Cycle

### DISCUSSION THREAD

**Discussion by daeynasvistas** [ngraziano/LMICPP-Arduino#28] [Q&A] [answered]
this is great..
ok.. so using void Lmic::setDutyRate(uint8_t rate) .. i can enforce duty cycle (regions)?

in multiple example i can see:
`LMIC.setDutyRate(12);`

apply this on my code:

i can see that the payload are "delayed" when apply.. so duty cycle did change and less payoad are sent .. or the interval between payload are increased.

![image](https://user-images.githubusercontent.com/2634610/169514738-87316135-901a-4296-9cb8-f4615612b205.png)

```
20519878 Uplink delayed until 35740282
20519897 Packet queued
20519906 sec TX_INTERVAL 3875000
24395875 Uplink delayed until 35740282
24395894 Packet queued
24395903 sec TX_INTERVAL 4625000
29020938 Uplink delayed until 35740282
29020956 Packet queued
29020965 sec TX_INTERVAL 5812500
34834937 Uplink delayed until 35740282
34834956 Packet queued
34834965 sec TX_INTERVAL 6687500
35740175 Ready for uplink
35740235 Build pkt # 2
35740246 Time on air : 66 ms
```

**can someone explain the "12" value?**

some theoretical explanation
https://www.thethingsnetwork.org/docs/lorawan/duty-cycle/

**Comment by ngraziano** [comment]
Hi 
LMIC.setDutyRate(xx); is equivalent of the network sending a MAC "DutyCycleReq" command to change the duty cycle of the device.
This duty cycle is global and is different than the one define by the regional setting.

The TTN community network have a fair use of 30 seconds per day (24 hours) per node.
It can be enforced by a duty cycle of 30/ (60*60*24) = 1 / 2880. But lmicpp only allow duty cycle define of type 1/2^dutyrate, the nearest suitable value is 1/4096 = 1/2^**12**.  This duty cycle is set in the example to prevent people to go beyond this fair use by accident.

On LMICPP you can only queue one packet, so only the last packet is sent when you hit duty cycle.

**Comment by daeynasvistas** [comment, marked as answer]
This is great.. im testing with ESP8266 (i also use ESP32 with i2c EEPROM) and try to save LoraState in "EEPROM" (emulated) so dont need to OTAA again when power goes off/on

if i use somethink like**Lmic::saveState(store); |  LMIC.loadState(retrieve);**  even if i didnt define LMIC.setDutyRate(xx);  i get the delay .

```
20519906 sec TX_INTERVAL 3875000
24395875 Uplink delayed until 35740282
```

i have behavior that i don't understand


Let say i flash ESP.. then first time .. LMICPP do OTAA, join request/accept and send first payload... node are now insync.
then i power off node .. power on .. the LMICPP read state from EEPROM  and try to send first payload 

```
Setup(){
  (...)
  nextSend = os_getTime() ;
}
```
![image](https://user-images.githubusercontent.com/2634610/169559703-9be051bc-9e27-4290-829b-8cfede140625.png)

From FCnt= 2 and FCnt= 3 i did power off/onn device.. so all GREAT!
but not allways ..s ometime.. without using LMIC.setDutyRate(xx);  

When i Power OFF/ON .. the First payload are delayed.. 
![image](https://user-images.githubusercontent.com/2634610/169576027-233273f6-0b1d-465b-86b2-1d480c6bf36a.png)


if i wait... it may send again. 
if it send again .. all the next send are not delayed.

then i power off/on node.. delay.. delay .. 

**is there some default LMIC.setDutyRate(xx); apply on the lib when its save/load lora state?**

is this one related?
![image](https://user-images.githubusercontent.com/2634610/169566057-63e8e0ef-fd79-4019-8b32-1f052c729bf3.png)

![image](https://user-images.githubusercontent.com/2634610/169574386-a110292c-ea8f-4ab5-89d7-ff262bf7ad16.png)

**Reply by ngraziano** [reply]
I think you problem is there is also a duty cycle for the regional  limitation and if os_getTime do not increase during sleep, there will be a problem. 

If os_getTime is reset to zero at start, at each startup it will wait to the next allowed time for a band. This duty cycle has nothing to do with  LMIC.setDutyRate, it hard coded to respect the legal limitation.

You can easily check by print the result of os_getTime().tick() and compare to a real clock (1 tick is equal to 16 microsecond)

There is two solution : 

1. Make sure to use RTC for os_getTime (I do it for ESP32 but i am not sure for ESP8266) and check that RTC continue to work during sleep.

1. Bypass the save of band utilisation by using saveStateWithoutTimeData and loadStateWithoutTimeData. But you must be sure to manually respect the limitation of your region (see https://www.thethingsnetwork.org/docs/lorawan/duty-cycle/#maximum-duty-cycle for exemple)

**Reply by daeynasvistas** [reply]
ok.. yes os_getTime() are probably zero, because i dont use RTC.I disconnect the controler from power. its not deep sleep.


> Make sure to use RTC for os_getTime (I do it for ESP32 but i am not sure for ESP8266) and check that RTC continue to work during sleep.

dont have external RTC in that node. thats why i need to save state in EEPROM

so  will check the other alternative.

**Comment by daeynasvistas** [comment]
Works very well!

Power ON device first time ever. 
Device makes OTAA join and then send first payload to sync with NS server.

after that.. it will send payload and apply LMIC.setDutyRate(xx);
im saving state every time i send payload 

if the device gets disconnected from source power .. when power ON it will send payload without OTAA. And without delay.. after that the duty rate use the LMIC.setDutyRate(xx); 

so,unless the devices gets in some poweroff poweron cycle, some reset loop (i need to do some change to prevente this) it respect the limitation of the region
